### PR TITLE
2547 fix broken tests

### PR DESCRIPTION
**Pull request by iilyak**

### CODE REVIEW

File: src/couch_proc_manager.erl [hunk 1]
```diff
@@ -449,13 +449,23 @@ return_proc(#state{} = State, #proc_int{} = ProcInt) ->
...
-                State
+                maybe_assign_proc(State, ProcInt)
         end;
     false ->
         remove_proc(State, ProcInt)
     end,
     flush_waiters(NewState, Lang).
 
+maybe_assign_proc(#state{} = State, ProcInt) ->
+    #proc_int{pid = Pid, lang = Lang} = ProcInt,
```
Comment (kxepal) [inline src/couch_proc_manager.erl]: `couch/src/couch_proc_manager.erl:460: Warning: variable 'Pid' is unused`

File: src/test_util.erl [hunk 2]
```diff
@@ -86,43 +67,46 @@ start_couch(ExtraApps) ->
...
+    start_applications(Apps, []).
 
-start_applications([]) ->
-    ok;
-start_applications([App|Apps]) ->
-    case application:start(App) of
+start_applications([], Acc) ->
+    lists:reverse(Acc);
+start_applications([App|Apps], Acc) ->
+    NewAcc =
```
Comment (kxepal) [inline src/test_util.erl]: `couch/src/test_util.erl:95: Warning: variable 'NewAcc' is unused`

File: src/couch_auth_cache.erl [hunk 3]
```diff
@@ -231,8 +232,18 @@ handle_info(restart_event_listener, State) ->
             ?MODULE, handle_db_event, nil, [{dbname, AuthDbName}]
         ),
     {noreply, State#state{event_listener=NewListener}};
+handle_info({'DOWN', Ref, _, _, shutdown}, State) ->
```
Comment (kxepal) [inline src/couch_auth_cache.erl]: `couch/src/couch_auth_cache.erl:235: Warning: variable 'Ref' is unused`

**Comment by kxepal** [comment]
Compaction tests doesn't feels well:
https://gist.github.com/kxepal/7ed999211a24f2744c52

**Comment by kxepal** [comment]
`couch/test/couch_task_status_tests.erl:96: Warning: function should_control_update_frequency/1 is unused` - it's actually commended in test suite.

File: test/couch_changes_tests.erl [hunk 4]
```diff
@@ -38,13 +35,18 @@ setup() ->
         save_doc(Db, {[{<<"_id">>, <<"doc5">>}]})
     ]],
     Rev = lists:nth(3, Revs),
-    {ok, Rev1} = save_doc(Db, {[{<<"_id">>, <<"doc3">>}, {<<"_rev">>, Rev}]}),
+    couch_db:ensure_full_commit(Db),
+    {ok, Db1} = couch_db:reopen(Db),
+
+    Doc = couch_db:open_doc(Db1, <<"doc3">>),
```
Comment (kxepal) [inline test/couch_changes_tests.erl]: `couch/test/couch_changes_tests.erl:41: Warning: variable 'Doc' is unused`

**Comment by kxepal** [comment]
Looks like missed in previous changes:
`couch/test/couch_stream_tests.erl:113: Warning: a term is constructed, but never used` - those assertions remain uncovered.

File: include/couch_eunit.hrl [hunk 5]
```diff
@@ -31,6 +31,11 @@
     filename:join([?BUILDDIR(), "src", "couch", "test", "fixtures"])).
 -define(TEMPDIR,
     filename:join([?BUILDDIR(), "tmp", "tmp_data"])).
+-define(ABS_PATH(File),
+    filename:join([
+        filename:dirname(
+            filename:dirname(
+                filename:dirname(filename:absname(File)))), File])).
```
Comment (kxepal) [inline include/couch_eunit.hrl]: I lost myself here counting what the actual path here will be. Is there way to make it better?

Comment (iilyak) [inline include/couch_eunit.hrl]: couch/test/couch_stream_tests.erl:113: 
Is a strange one

**Comment by iilyak** [comment]
Test `should_control_update_frequency` is commented out because the functionality it tests is disabled https://github.com/apache/couchdb-couch/commit/5d1334028123ae3002a1c7c8d62de8da021bc3cd

**Comment by iilyak** [comment]
Hmm, I cannot reproduce this https://gist.github.com/kxepal/7ed999211a24f2744c52

**Comment by kxepal** [comment]
It seems some race condition happened for those run. Stripping any other tests and run only compaction one gives me constantly "one test canceled" and failure in the result.

https://gist.github.com/kxepal/40f5a6dd4ad13af000ac

**Comment by kxepal** [comment]
Oh, there timeout happened. It's hard to notice the real problem in this logging spam /:

**Comment by iilyak** [comment]
I agree the noise level is horrible. I would love to see only `<test>...ok` by default and also have `-v` or something. So the one can enable logs for debugging failing test cases. But it is not easy with eunit.

**Comment by kxepal** [comment]
@iilyak well, that's how eunit worked for in 1963-eunit branch on apache/couchdb repo. We just need to disable sasl and backport lager control code from rcouch.

**Comment by iilyak** [comment]
Thank you @kxepal for pointing out to timeout. Here is what happens:
- [default timeout is 5 sec](https://github.com/erlang/otp/blob/maint/lib/eunit/src/eunit_internal.hrl#L12) 
- [we wait at least 6 sec](https://github.com/iilyak/couchdb-couch/blob/2547-fix-broken-tests/test/couchdb_compaction_daemon.erl#L77-L81)

**Comment by kxepal** [comment]
@iilyak but `{timeout, ?TIMEOUT_S` instructs to wait for 30 seconds, right? Or so it should...

**Comment by iilyak** [comment]
@kxepal Yes you are right `{timeout, ?TIMEOUT_S` works as expected. Could you try increasing the value of `TIMEOUT_S` to 60 sec? If it would work I'll push the update to the PR.

**Comment by kxepal** [comment]
@iilyak cannot check since it starts fails for me with enoent error as previous report was. But I got the reason:
```
20:28:29.004 [error] gen_server couch_compaction_daemon terminated with reason: {compaction_loop_died,{{badmatch,{error,enoent}},[{couch_compaction_daemon,abs_path,1},{couch_compaction_daemon,free_space_rec,2},{couch_compaction_daemon,can_db_compact,2},{couch_compaction_daemon,maybe_compact_db,2},{couch_compaction_daemon,'-compact_loop/1-fun-0-',2},{couch_server,'-all_databases/2-fun-0-',4},{filelib,do_fold_files2,8},{couch_server,all_databases,2}]}}
```

**Comment by kxepal** [comment]
Ok, finally got the situation when test had passed. 60 seconds is more than enough - approximately it needs ~40-50 seconds for all the things.

However, two notices:

1. It doesn't exits gracefully:

```
02:53:15.709 [error] CRASH REPORT Process <0.1400.0> with 0 neighbours exited with reason: killed in gen_server:terminate/7 line 804
02:53:15.711 [error] CRASH REPORT Process <0.1399.0> with 0 neighbours exited with reason: killed in gen_server:terminate/7 line 804
02:53:15.711 [error] gen_server <0.1419.0> terminated with reason: killed
02:53:15.712 [error] CRASH REPORT Process <0.1419.0> with 0 neighbours exited with reason: killed in gen_server:terminate/7 line 804
02:53:15.712 [error] gen_server <0.1418.0> terminated with reason: killed
02:53:15.712 [error] gen_server <0.1417.0> terminated with reason: killed
02:53:15.714 [error] CRASH REPORT Process <0.1418.0> with 0 neighbours exited with reason: killed in gen_server:terminate/7 line 804
02:53:15.714 [error] gen_server <0.1411.0> terminated with reason: killed
02:53:15.715 [error] CRASH REPORT Process <0.1417.0> with 0 neighbours exited with reason: killed in gen_server:terminate/7 line 804
02:53:15.716 [error] CRASH REPORT Process <0.1411.0> with 0 neighbours exited with reason: killed in gen_server:terminate/7 line 804
02:53:15.716 [error] gen_server <0.1410.0> terminated with reason: killed
02:53:15.716 [error] gen_server <0.1389.0> terminated with reason: killed
02:53:15.716 [error] gen_server <0.1388.0> terminated with reason: killed
02:53:15.717 [error] CRASH REPORT Process <0.1389.0> with 0 neighbours exited with reason: killed in gen_server:terminate/7 line 804
02:53:15.718 [error] CRASH REPORT Process <0.1410.0> with 0 neighbours exited with reason: killed in gen_server:terminate/7 line 804
02:53:15.718 [error] gen_server <0.1438.0> terminated with reason: killed
02:53:15.719 [error] CRASH REPORT Process <0.1388.0> with 0 neighbours exited with reason: killed in gen_server:terminate/7 line 804
02:53:15.719 [error] gen_server <0.1437.0> terminated with reason: killed
02:53:15.720 [error] CRASH REPORT Process <0.1438.0> with 0 neighbours exited with reason: killed in gen_server:terminate/7 line 804
```

2. Why it have to run for so long? Previously 30 seconds was more than enough to done all the things on slow and constraint with resources virtual box. Here I cannot fit the limit with my real notebook /:

**Comment by robertkowalski** [comment]
My system: OSX with Erlang/OTP 17.4 

```
module 'couchdb_modules_load_tests'
  Verify that all modules loads
    couchdb_modules_load_tests:62: should_load_module (couch_auth_cache)...[0.002 s] ok
    couchdb_modules_load_tests:62: should_load_module (couch_btree)...[0.003 s] ok
    couchdb_modules_load_tests:62: should_load_module (couch_changes)...[0.003 s] ok
    couchdb_modules_load_tests:62: should_load_module (couch_compress)...[0.002 s] ok
    couchdb_modules_load_tests:62: should_load_module (config)...*failed*
in function couchdb_modules_load_tests:'-should_load_module/1-fun-0-'/1 (test/couchdb_modules_load_tests.erl, line 62)
**error:{assertMatch_failed,[{module,couchdb_modules_load_tests},
                     {line,62},
                     {expression,"code : load_file ( Mod )"},
                     {pattern,"{ module , _ }"},
                     {value,{error,not_purged}}]}
00:22:10.137 [error] Loading of /Users/robert/apache/couchdb/src/config/ebin/config.beam failed: not_purged


    couchdb_modules_load_tests:62: should_load_module (config_writer)...[0.009 s] ok
```

**Comment by kxepal** [comment]
`not_purged` [means|http://www.erlang.org/doc/man/erlang.html#load_module-2]:
> Binary contains a module which cannot be loaded because old code for this module already exists.

Run make clean to drop old beam files.

**Comment by robertkowalski** [comment]
oh ok!

now i get:

```
module 'couch_work_queue'
  module 'couch_work_queue_tests'
    Single producer and consumer
      Queue with 3 max items
        couch_work_queue_tests:143: should_have_no_items_for_new_queue...ok
        undefined
        *** instantiation of subtests failed ***
**in function couch_work_queue_tests:'-should_block_producer_on_full_queue_count/1-fun-1-'/2 (test/couch_work_queue_tests.erl, line 206)
in call from couch_work_queue_tests:should_block_producer_on_full_queue_count/1 (test/couch_work_queue_tests.erl, line 206)
**error:{assertEqual_failed,[{module,couch_work_queue_tests},
                     {line,206},
                     {expression,"couch_work_queue : item_count ( Q )"},
                     {expected,2},
                     {value,1}]}
```

**Comment by iilyak** [comment]
@robertkowalski Could you test it again? I pushed the fix.

**Comment by robertkowalski** [comment]
@iilyak works now! thanks!

**Comment by sebastianrothbucher** [comment]
looks good to me also - just a little catch for the javascript tests: they require /_utils/script/test/lorem(_b64).txt in place and fail if the files are not there (at least one: attachments.js). Quick remedy is copying the files obviously, but that's not good. Anyone else having the same?

**Comment by iilyak** [comment]
@sebastianrothbucher I didn't work on JS test suite in the PR. I've got an impression that someone else is working on this.

**Comment by sebastianrothbucher** [comment]
@iilyak sorry, just realized it's different projects - did a PR there. eunit looks good as far as I can tell... - great!

**Comment by kxepal** [comment]
@iilyak thank you a lot!

**Pull request merged**